### PR TITLE
fix test of presence of attack/second_attack in filter_attack function

### PR DESCRIPTION
**Pull request by newfrenchy83**
resolve problem in https://github.com/wesnoth/wesnoth/pull/7509#discussion_r1165220995discussion

### CODE REVIEW

**Comment by newfrenchy83** [comment]
@stevecotton @CelticMinstrel i hope what this fix is conveniant for bothof you, ah, i also repulll old declaration for op_ctx because abuse of auto declaration is not readable.

File: src/game_events/handlers.cpp [hunk 1]
```diff
@@ -247,18 +247,23 @@ struct filter_attack : public event_filter {
...
-				const_attack_ptr second_attack = std::make_shared<const attack_type>(*temp_other_weapon);
-				auto ctx = attack->specials_context(u, opp, loc, loc_d, first_, second_attack);
-				auto opp_ctx = second_attack->specials_context(opp, u, loc_d, loc, !first_, attack);
-				return swf_.empty() || attack->matches_filter(swf_.get_parsed_config());
-			} else {
-				auto ctx = attack->specials_context(u, loc, first_);
-				return swf_.empty() || attack->matches_filter(swf_.get_parsed_config());
+			auto temp_weapon = event_info.data.mandatory_child(first_ ? "first" : "second");
+			const_attack_ptr attack = std::make_shared<const attack_type>(temp_weapon);
+			if(attack){
```
Comment (stevecotton) [inline src/game_events/handlers.cpp:252]: `attack` can never be `nullptr` here, so this seems to be testing the wrong condition.

File: src/game_events/handlers.cpp [hunk 2]
```diff
@@ -247,18 +247,23 @@ struct filter_attack : public event_filter {
...
+				if(unit_d != units.end() && loc_d.matches_unit(unit_d)) {
+					const auto opp = unit_d->shared_from_this();
+					auto temp_other_weapon = event_info.data.mandatory_child(!first_ ? "first" : "second");
```
Comment (stevecotton) [inline src/game_events/handlers.cpp:255]: Are "first" and "second" both mandatory, even if the defender doesn't have a weapon? I'm testing that at the moment.

File: src/game_events/handlers.cpp [hunk 3]
```diff
@@ -247,18 +247,23 @@ struct filter_attack : public event_filter {
...
+					const_attack_ptr second_attack = std::make_shared<const attack_type>(temp_other_weapon);
+					auto ctx = attack->specials_context(u, opp, loc, loc_d, first_, second_attack);
+					std::optional<decltype(ctx)> opp_ctx;
```
Comment (stevecotton) [inline src/game_events/handlers.cpp:258]: Seems less readable than `auto opp_ctx =` to me.

File: src/game_events/handlers.cpp (hunk 1 above)
Comment (CelticMinstrel) [reply src/game_events/handlers.cpp:252]: And you're using `mandatory_child` which means `temp_weapon` will never be null either. (If the child doesn't exist, an exception is thrown.)

Comment (CelticMinstrel) [reply src/game_events/handlers.cpp:252]: By the way, does this crash if you put `[filter_attack]` in a moveto event (for example)?

File: src/game_events/handlers.cpp (hunk 2 above)
Comment (newfrenchy83) [reply src/game_events/handlers.cpp:255]: what you suggest, in the end?

File: src/game_events/handlers.cpp (hunk 1 above)
Comment (newfrenchy83) [reply src/game_events/handlers.cpp:252]: must i remove if(attack) loop or could i let like it?

File: src/game_events/handlers.cpp (hunk 2 above)
Comment (newfrenchy83) [reply src/game_events/handlers.cpp]: reuse optionnal_child and added check attack/second_atack

File: src/game_events/handlers.cpp (hunk 3 above)
Comment (CelticMinstrel) [reply src/game_events/handlers.cpp:258]: I don't think `auto opp_ctx` works? That requires you to initialize it immediately, but the whole point is that we may or may not want to initialize it.

**Comment by CelticMinstrel** [comment]
I think this fix looks fine.

**Comment by newfrenchy83** [comment]
> I think this fix looks fine.

Ok

File: src/game_events/handlers.cpp (hunk 1 above)
Comment (stevecotton) [reply src/game_events/handlers.cpp]: Tested, and it now works okay even in a moveto event.

**Pull request merged**